### PR TITLE
Initial pom, README, etc

### PR DESCRIPTION
**Pull request by wolf31o2**
This is the initial set of files necessary to get a build going.

- [x] LICENSE.txt
- [x] README.md
- [x] pom.xml

This was tested using:
```
mvn clean package -DskipTests -P examples,templates,dist,release,rpm-prepare,rpm,deb-prepare,deb,tgz,unit-tests -Dgpg.passphrase=${GPG_PASSPHRASE} -Dgpg.useagent=false
```

### CODE REVIEW

**Comment by chtyim** [comment]
You don't need the checkstyle files, do you?

**Comment by wolf31o2** [comment]
Once Maven complained about the `suppressions.xml` I just copied both. However, I just tested and it looks like I don't need it. Removing.

**Comment by wolf31o2** [comment]
Once Maven complained about the `suppressions.xml` I just copied both. However, I just tested and it looks like I don't need it. Removing.

**Comment by anwar6953** [comment]
Can this go under release-tools repo?

**Comment by wolf31o2** [comment]
No. This is public and will be used by our build system for building the CDAP release packages, per discussion with @sreevatsanraman earlier today.

**Comment by chtyim** [comment]
What maven complains about the `suppressions.xml`?? It for checkstyle only and I think each submodule repo has their own checkstyle files.

**Comment by wolf31o2** [comment]
Each submodule repo does have its own files, however, the Maven plugin doesn't work like that. It takes the path based on where the command was run.

```
$ rm -f suppressions.xml
$ mvn clean package -DskipTests
[INFO] Scanning for projects...
[WARNING]
[WARNING] Some problems were encountered while building the effective model for co.cask.cdap:cdap-kafka:jar:3.5.0-SNAPSHOT
[WARNING] The expression ${build.directory} is deprecated. Please use ${project.build.directory} instead.
[WARNING] The expression ${build.directory} is deprecated. Please use ${project.build.directory} instead.
[WARNING]
[WARNING] Some problems were encountered while building the effective model for co.cask.cdap:cdap-data-fabric:jar:3.5.0-SNAPSHOT
[WARNING] 'dependencies.dependency.(groupId:artifactId:type:classifier)' must be unique: co.cask.cdap:cdap-common:test-jar -> duplicate declaration of version ${project.version} @ line 46, column 17
[WARNING]
[WARNING] Some problems were encountered while building the effective model for co.cask.cdap:cdap-spark-core:jar:3.5.0-SNAPSHOT
[WARNING] The expression ${build.directory} is deprecated. Please use ${project.build.directory} instead.
[WARNING] The expression ${build.directory} is deprecated. Please use ${project.build.directory} instead.
[WARNING] The expression ${build.directory} is deprecated. Please use ${project.build.directory} instead.
[WARNING]
[WARNING] It is highly recommended to fix these problems because they threaten the stability of your build.
[WARNING]
[WARNING] For this reason, future Maven versions might no longer support building such malformed projects.
[WARNING]
[INFO] Inspecting build with total of 51 modules...
[INFO] Installing Nexus Staging features:
[INFO]   ... total of 50 executions of maven-deploy-plugin replaced with nexus-staging-maven-plugin
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO]
[INFO] Cask Data Application Platform (CDAP)
[INFO] CDAP API Common
[INFO] CDAP SPI
[INFO] CDAP API
[INFO] CDAP Watchdog API
[INFO] CDAP Protocol
[INFO] CDAP Security SPI
[INFO] CDAP Build Common
[INFO] CDAP Common
[INFO] CDAP Data Format
[INFO] CDAP Notifications APIs
[INFO] CDAP Security
[INFO] CDAP Explore Client
[INFO] CDAP Data-Fabric
[INFO] CDAP Watchdog
[INFO] CDAP Notifications
[INFO] CDAP HBase 0.98 Compatibility
[INFO] CDAP Data-Fabric Tests
[INFO] CDAP Spark API
[INFO] CDAP Java Client
[INFO] CDAP Test Framework
[INFO] CDAP App-Fabric
[INFO] CDAP Explore JDBC Driver
[INFO] CDAP Explore
[INFO] CDAP Spark Core
[INFO] CDAP Unit Test Framework
[INFO] CDAP Kafka Server
[INFO] CDAP CLI
[INFO] CDAP Gateway
[INFO] CDAP UI
[INFO] CDAP Standalone
[INFO] CDAP Java Client Tests
[INFO] CDAP CLI Tests
[INFO] CDAP App Fabric Tests
[INFO] CDAP HBase 0.96 Compatibility
[INFO] CDAP HBase 1.0-CDH5.5.0 Compatibility
[INFO] CDAP HBase 1.0-CDH Compatibility
[INFO] CDAP HBase 1.0 Compatibility
[INFO] CDAP HBase 1.1 Compatibility
[INFO] CDAP HBase 1.2-CDH5.7.0 Compatibility
[INFO] CDAP Documentation Generator
[INFO] CDAP KMS
[INFO] CDAP Master
[INFO] CDAP Distributions
[INFO] CDAP Application Archetypes
[INFO] CDAP Spark Java Archetype
[INFO] CDAP Spark Scala Archetype
[INFO] CDAP MapReduce Archetype
[INFO] CDAP Application Archetype
[INFO] CDAP Integration Test Framework
[INFO] Cask Data Application Platform (CDAP) build
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building Cask Data Application Platform (CDAP) 3.5.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ cdap ---
[INFO] Deleting /home/chris/cdap-build/cdap/target
[INFO]
[INFO] --- apache-rat-plugin:0.10:check (rat-check) @ cdap ---
[INFO] 96 implicit excludes (use -debug for more details).
[INFO] Exclude: build-number.txt
[INFO] Exclude: LICENSE*.txt
[INFO] Exclude: *.rst
[INFO] Exclude: *.md
[INFO] Exclude: **/*.md
[INFO] Exclude: logs/**
[INFO] Exclude: cdap-docs/**
[INFO] Exclude: **/grok/patterns/**
[INFO] Exclude: conf/**
[INFO] Exclude: data/**
[INFO] Exclude: plugins/**
[INFO] Exclude: **/Apache-LICENSE-2.0
[INFO] Exclude: **/VERSION
[INFO] Exclude: **/*.patch
[INFO] Exclude: **/logrotate.d/**
[INFO] Exclude: **/limits.d/**
[INFO] Exclude: **/*.avsc
[INFO] Exclude: cdap-examples/**
[INFO] Exclude: cdap-app-templates/**
[INFO] Exclude: **/*.json
[INFO] Exclude: **/resources/**/*.properties
[INFO] Exclude: **/*.json.template
[INFO] Exclude: **/MANIFEST.MF
[INFO] Exclude: **/.rubocop*.yml
[INFO] 5 resources included (use -debug for more details)
[INFO] Rat check: Summary of files. Unapproved: 0 unknown: 0 generated: 0 approved: 4 licence.
[INFO]
[INFO] --- maven-checkstyle-plugin:2.12.1:check (validate) @ cdap ---
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] Cask Data Application Platform (CDAP) ............. FAILURE [9.834s]
[INFO] CDAP API Common ................................... SKIPPED
[INFO] CDAP SPI .......................................... SKIPPED
[INFO] CDAP API .......................................... SKIPPED
[INFO] CDAP Watchdog API ................................. SKIPPED
[INFO] CDAP Protocol ..................................... SKIPPED
[INFO] CDAP Security SPI ................................. SKIPPED
[INFO] CDAP Build Common ................................. SKIPPED
[INFO] CDAP Common ....................................... SKIPPED
[INFO] CDAP Data Format .................................. SKIPPED
[INFO] CDAP Notifications APIs ........................... SKIPPED
[INFO] CDAP Security ..................................... SKIPPED
[INFO] CDAP Explore Client ............................... SKIPPED
[INFO] CDAP Data-Fabric .................................. SKIPPED
[INFO] CDAP Watchdog ..................................... SKIPPED
[INFO] CDAP Notifications ................................ SKIPPED
[INFO] CDAP HBase 0.98 Compatibility ..................... SKIPPED
[INFO] CDAP Data-Fabric Tests ............................ SKIPPED
[INFO] CDAP Spark API .................................... SKIPPED
[INFO] CDAP Java Client .................................. SKIPPED
[INFO] CDAP Test Framework ............................... SKIPPED
[INFO] CDAP App-Fabric ................................... SKIPPED
[INFO] CDAP Explore JDBC Driver .......................... SKIPPED
[INFO] CDAP Explore ...................................... SKIPPED
[INFO] CDAP Spark Core ................................... SKIPPED
[INFO] CDAP Unit Test Framework .......................... SKIPPED
[INFO] CDAP Kafka Server ................................. SKIPPED
[INFO] CDAP CLI .......................................... SKIPPED
[INFO] CDAP Gateway ...................................... SKIPPED
[INFO] CDAP UI ........................................... SKIPPED
[INFO] CDAP Standalone ................................... SKIPPED
[INFO] CDAP Java Client Tests ............................ SKIPPED
[INFO] CDAP CLI Tests .................................... SKIPPED
[INFO] CDAP App Fabric Tests ............................. SKIPPED
[INFO] CDAP HBase 0.96 Compatibility ..................... SKIPPED
[INFO] CDAP HBase 1.0-CDH5.5.0 Compatibility ............. SKIPPED
[INFO] CDAP HBase 1.0-CDH Compatibility .................. SKIPPED
[INFO] CDAP HBase 1.0 Compatibility ...................... SKIPPED
[INFO] CDAP HBase 1.1 Compatibility ...................... SKIPPED
[INFO] CDAP HBase 1.2-CDH5.7.0 Compatibility ............. SKIPPED
[INFO] CDAP Documentation Generator ...................... SKIPPED
[INFO] CDAP KMS .......................................... SKIPPED
[INFO] CDAP Master ....................................... SKIPPED
[INFO] CDAP Distributions ................................ SKIPPED
[INFO] CDAP Application Archetypes ....................... SKIPPED
[INFO] CDAP Spark Java Archetype ......................... SKIPPED
[INFO] CDAP Spark Scala Archetype ........................ SKIPPED
[INFO] CDAP MapReduce Archetype .......................... SKIPPED
[INFO] CDAP Application Archetype ........................ SKIPPED
[INFO] CDAP Integration Test Framework ................... SKIPPED
[INFO] Cask Data Application Platform (CDAP) build ....... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 14.361s
[INFO] Finished at: Tue Aug 02 23:50:18 UTC 2016
[INFO] Final Memory: 42M/669M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:2.12.1:check (validate) on project cdap: Failed during checkstyle configuration: cannot initialize module SuppressionFilter - Cannot set property 'file' in module SuppressionFilter to 'suppressions.xml': unable to find suppressions.xml: InvocationTargetException -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

I tried setting the file in the cdap pom to ./suppressions.xml and that didn't work, either.

**Comment by anwar6953** [comment]
Can we execute the mvn command from the actual subdirectory, so it uses the repo's actual suppressions.xml?
The current approach doesn't allow different sub-directories to have different suppressions.xml.
It also doesn't honor the repo's suppressions.xml, right?

**Comment by wolf31o2** [comment]
No. We're trying to solve a specific problem. That problem is that our build system currently needs to be aware of which repository it is building. This means we have to constantly fork our builds, as new external repositories are added. This must be done by hand, is error prone, time consuming, and it takes the important information of how to build our software out of the repositories and into a system that is not easily exported or transferable.

Running the Maven commands from the individual subdirectories is what we do now. It requires manual work every release to maintain, and puts additional strain on our build system resources. It also requires updating our internal `cdap_auto` cookbook every release.

The idea is to have a single build repository which captures all of the build information in one place, making it easily to replicate for both developers and our builds. I was hoping to let Maven be that tool. If that won't work, I will need a separate tool to codify the build information.

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>co.cask.cdap.build</groupId>
+  <artifactId>cdap-build</artifactId>
+  <version>1.0</version>
+  <packaging>pom</packaging>
+  <name>Cask Data Application Platform (CDAP) build</name>
+
+  <modules>
+    <module>cdap</module>
```
Comment (anwar6953) [inline pom.xml]: so there's going to be a module for each repo?
And that module will be created during runtime of the tool?

Comment (wolf31o2) [inline pom.xml]: Correct. Branches for each release will include the Maven modules supported as well as the Git submodules required to do so. I am only including `cdap` here as it is the only module/repository that applies to all `cdap` release branches. After this is merged, I will be creating branches for `release/2.8` through `release/3.5` and including the correct repositories/modules in those PR.

`release/3.3`+ get `hydrator-plugins` tracking the appropriate release branch in that repo
`release/3.4`+ get `cask-tracker` and `cdap-navigator` added to the list
`release/3.5`+ get `cdap-security-extn` added to the list

Comment (anwar6953) [inline pom.xml]: not a big deal, but wonder if we should remove branch-specific repos from the general `README.md`.

Comment (wolf31o2) [inline pom.xml]: We should. I'll do that when I do the release branches in this repo, unless you think that I should go ahead and do it. I'll have to add them back to develop, anyway, once I add all of the Git submodules for them.

**Comment by chtyim** [comment]
It's due to the cdap checkstyle.xml is not right (it shouldn't have the SuppressionFilter inside, but instead should relies on the checkstyle plugin config to do that). Anyway, I'll open a PR in CDAP to get that fixed (the same fix need to applies to all other repos).

**Comment by wolf31o2** [comment]
Removed suppressions.xml awaiting PR from @chtyim

**Comment by wolf31o2** [comment]
PR: caskdata/cdap#6329

**Comment by chtyim** [comment]
LGTM

**Pull request merged**